### ISSUE TITLE
Questions about gradient propagation in code , i am a new hand.

### ISSUE THREAD

**Issue by ho95ly** [jiecaoyu/XNOR-Net-PyTorch#49]
I am a novice and get confused about gradient propagation in this code about training part.
So i have some problems , these may be "stupid".

1. Can you tell me the gradient propagation how it works..I don`t understand the meaning about:
      **# restore weights
        bin_op.restore()
        bin_op.updateBinaryGradWeight()**
It seems after backward, we update weights again? but why need restore? 
the original code shows below.

```
def train(epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        if args.cuda:
            data, target = data.cuda(), target.cuda()
        data, target = Variable(data), Variable(target)
        optimizer.zero_grad()

        # process the weights including binarization
        bin_op.binarization()

        output = model(data)
        loss = criterion(output, target)
    
        loss.backward()

        # restore weights
        bin_op.restore()
        bin_op.updateBinaryGradWeight()

        optimizer.step()
        if batch_idx % args.log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.item()))
    return

```
2. Another problem shows in code below.  I don not understand these two parts.
   **m.weight.data.zero_().add_(1.0)** and   **m.weight.data.clamp_(min=0.01)**
 It seems BatchNorm1d (or 2d) layer`s weight should set some options during initializing and forward?
```

class LeNet_5(nn.Module):
    def __init__(self):
        super(LeNet_5, self).__init__()
        self.conv1 = nn.Conv2d(1, 20, kernel_size=5, stride=1)
        self.bn_conv1 = nn.BatchNorm2d(20, eps=1e-4, momentum=0.1, affine=False)
        self.relu_conv1 = nn.ReLU(inplace=True)
        self.pool1 = nn.MaxPool2d(kernel_size=2, stride=2)
        self.bin_conv2 = BinConv2d(20, 50, kernel_size=5, stride=1, padding=0)
        self.pool2 = nn.MaxPool2d(kernel_size=2, stride=2)
        self.bin_ip1 = BinConv2d(50*4*4, 500, Linear=True,
                previous_conv=True, size=4*4)
        self.ip2 = nn.Linear(500, 10)

        for m in self.modules():
            if isinstance(m, nn.BatchNorm2d) or isinstance(m, nn.BatchNorm1d):
                if hasattr(m.weight, 'data'):
                    m.weight.data.zero_().add_(1.0)
        return

    def forward(self, x):
        for m in self.modules():
            if isinstance(m, nn.BatchNorm2d) or isinstance(m, nn.BatchNorm1d):
                if hasattr(m.weight, 'data'):
                    m.weight.data.clamp_(min=0.01)
        x = self.conv1(x)
        x = self.bn_conv1(x)
        x = self.relu_conv1(x)
        x = self.pool1(x)
        x = self.bin_conv2(x)
        x = self.pool2(x)

        # x = x.view(x.size(0), 50*4*4)

        x = self.bin_ip1(x)
        x = self.ip2(x)
        return x

```
Sincerely
THANK YOU VERY MUCH.

**Comment by jiecaoyu** [comment]
1. Since current weights in the network are binary. We need to restore the floating-point weights before applying the gradients.

2. BatchNorm weights are randomly initialized. Therefore, I fix the initialization to 1.0. ```m.weight.data.clamp_(min=0.01)``` force the weights to be larger than 0, which makes it easier to merge the BatchNorm with the previous Conv to save computation.

**Comment by ho95ly** [comment]
I just got into contact with binarization and I don’t understand much about the details. i will read the paper carefully.
But, thanks a lof for your reply.You really help me.
